dynamic_programming/fibonacci_numbers.nim

```diff
@@ -0,0 +1,186 @@
+## Fibonacci Numbers
+## ===========================================================================
+## Fibonacci numbers are numbers that are part of the Fibonacci sequence.
+## The Fibonacci sequence generally starts with 0 and 1, every next number
+## is the sum of the two preceding numbers:
+##   - F(0) = 0;
+##   - F(1) = 1;
+##   - F(n) = F(n-1) + F(n-2);
+##
+##  References:
+##    - https://en.wikipedia.org/wiki/Fibonacci_sequence
+##    - https://oeis.org/A000045
+{.push raises: [].}
+
+runnableExamples:
+
+  doAssert fibonacciSeqIterative(5) == @[Natural(0), 1, 1, 2, 3]
+
+  doAssert fibonacciClosure(4) == Natural(3)
+
+  for f in fibonacciIterator(1.Natural..4.Natural):
+    echo f # 1, 1, 2, 3
+
+import std/math
+
+type
+  fibMatrix = array[2, array[2, int]]
+
+func `*`(m1, m2: fibMatrix): fibMatrix =
+  let
+    a = m1[0][0] * m2[0][0] + m1[0][1] * m2[1][0]
+    b = m1[0][0] * m2[0][1] + m1[0][1] * m2[1][1]
+    z = m1[1][0] * m2[0][0] + m1[1][1] * m2[1][0]
+    y = m1[1][0] * m2[0][1] + m1[1][1] * m2[1][1]
```

Comment:
Why did you unroll the for loop here?

Reply:
it looks a lot more clearer to me than:
```Nim
  for i in 0..1:
    for j in 0..1:
      for k in 0..1:
        result[i][j] += m1[i][k] * m2[k][j]
```
And is quite a bit faster, at least in micro benchmarks.